aya-obj/src/btf/relocation.rs

```diff
@@ -256,16 +263,41 @@ impl Object {
     }
 }
 
-fn relocate_btf_program<'target>(
-    program: &mut Program,
+fn get_function_by_relocation<'a>(
+    section_index: &SectionIndex,
+    functions: &'a mut BTreeMap<(usize, u64), Function>,
+    rel: &Relocation,
+) -> Option<&'a mut Function> {
+    for (_, func) in functions.range_mut((
+        Included(&(section_index.0, 0)),
+        Included(&(section_index.0, u64::MAX)),
+    )) {
+        let ins_offset = rel.ins_offset / mem::size_of::<bpf_insn>();
+        let func_offset = func.section_offset / mem::size_of::<bpf_insn>();
+        let func_size = func.instructions.len();
+
+        if ins_offset >= func_offset && ins_offset < func_offset + func_size {
+            return Some(func);
+        }
+    }
+
+    None
+}
+
+fn relocate_btf_functions<'target>(
+    section_index: &SectionIndex,
+    functions: &mut BTreeMap<(usize, u64), Function>,
     relos: &[Relocation],
     local_btf: &Btf,
     target_btf: &'target Btf,
     candidates_cache: &mut HashMap<u32, Vec<Candidate<'target>>>,
 ) -> Result<(), RelocationError> {
     for rel in relos {
-        let instructions = &mut program.function.instructions;
-        let ins_index = rel.ins_offset / mem::size_of::<bpf_insn>();
+        let function = get_function_by_relocation(section_index, functions, rel)
```

Comment:
This works, but algorithmically it's pretty terrible. I think you should cache
the last Function and only lookup if the relocation is outside the last
function.